File: backend/src/main/java/com/isg/backend/violation/application/notification/AlertMessage.java
package com.isg.backend.violation.application.notification;

import com.isg.backend.violation.domain.ViolationType;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

public record AlertMessage(
        UUID eventId,
        long version,
        UUID violationId,
        ViolationType type,
        String cameraName,
        String departmentName,
        Instant startedAt,
        double confidence,
        String lifecycleStatus,
        String recordingStatus,
        boolean clipReady,
        boolean coverImageReady
) {

    public AlertMessage {
        Objects.requireNonNull(
                violationId,
                "violationId must not be null"
        );

        Objects.requireNonNull(
                eventId,
                "eventId must not be null"
        );

        Objects.requireNonNull(
                type,
                "type must not be null"
        );

        if (version < 0) {
            throw new IllegalArgumentException(
                    "version must not be negative"
            );
        }

        if (cameraName == null || cameraName.isBlank()) {
            throw new IllegalArgumentException(
                    "cameraName must not be blank"
            );
        }

        if (departmentName == null || departmentName.isBlank()) {
            throw new IllegalArgumentException(
                    "departmentName must not be blank"
            );
        }

        Objects.requireNonNull(
                startedAt,
                "startedAt must not be null"
        );

        if (confidence < 0.0 || confidence > 1.0) {
            throw new IllegalArgumentException(
                    "confidence must be between 0 and 1"
            );
        }

        if (lifecycleStatus == null || lifecycleStatus.isBlank()) {
            throw new IllegalArgumentException(
                    "lifecycleStatus must not be blank"
            );
        }

        if (recordingStatus == null || recordingStatus.isBlank()) {
            throw new IllegalArgumentException(
                    "recordingStatus must not be blank"
            );
        }
    }
}

File: mobile/lib/core/models/alert_message.dart
/// Backend `AlertMessage` (STOMP `/user/queue/alerts`).
///
/// Alanlar backend record'u ile birebirdir; `eventId`/`version` sözleşmede
/// yoktur ve uydurulmaz.
class AlertMessage {
  final String violationId;
  final String type;
  final String cameraName;
  final String departmentName;
  final DateTime startedAt;
  final double confidence;
  final String lifecycleStatus;
  final String recordingStatus;
  final bool clipReady;
  final bool coverImageReady;

  const AlertMessage({
    required this.violationId,
    required this.type,
    required this.cameraName,
    required this.departmentName,
    required this.startedAt,
    required this.confidence,
    required this.lifecycleStatus,
    required this.recordingStatus,
    required this.clipReady,
    required this.coverImageReady,
  });

  /// Zorunlu alan eksik/bozuksa [FormatException] atar; çağıran taraf bunu
  /// güvenli parse failure'a çevirir.
  factory AlertMessage.fromJson(Map<String, dynamic> json) {
    return AlertMessage(
      violationId: _requiredString(json['violationId'], 'violationId'),
      type: _requiredString(json['type'], 'type'),
      cameraName: _requiredString(json['cameraName'], 'cameraName'),
      departmentName: _requiredString(json['departmentName'], 'departmentName'),
      startedAt: _requiredInstant(json['startedAt'], 'startedAt'),
      confidence: _requiredDouble(json['confidence'], 'confidence'),
      lifecycleStatus:
          _requiredString(json['lifecycleStatus'], 'lifecycleStatus'),
      recordingStatus:
          _requiredString(json['recordingStatus'], 'recordingStatus'),
      clipReady: _requiredBool(json['clipReady'], 'clipReady'),
      coverImageReady:
          _requiredBool(json['coverImageReady'], 'coverImageReady'),
    );
  }

  static String _requiredString(Object? value, String field) {
    if (value is String && value.isNotEmpty) {
      return value;
    }
    throw FormatException('$field missing');
  }

  static DateTime _requiredInstant(Object? value, String field) {
    if (value is String) {
      final parsed = DateTime.tryParse(value);
      if (parsed != null) {
        return parsed;
      }
    }
    throw FormatException('$field missing');
  }

  static double _requiredDouble(Object? value, String field) {
    if (value is num) {
      return value.toDouble();
    }
    throw FormatException('$field missing');
  }

  static bool _requiredBool(Object? value, String field) {
    if (value is bool) {
      return value;
    }
    throw FormatException('$field missing');
  }
}
